### ISSUE TITLE
Rectangular training/inference results in a square model

### ISSUE THREAD

**Issue by Shassk** [ultralytics/ultralytics#7690]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I'm trying to train a model with a rectangular input tensor. From some issues' discussions it seems like this should be possible, however when I'm trying to run training in a Google Colab notebook like this:
```
from ultralytics import YOLO
model = YOLO('yolov8n.yaml')  # build a new model from YAML
model.train(data='/content/datasets/car-old/custom.yaml', epochs=100, rect=True, imgsz=[512,288], batch=56, cache=True)
```
I'm getting a square input tensor regardless:
![Снимок](https://github.com/ultralytics/ultralytics/assets/62873560/89ad582b-e75f-4867-98b3-34d533d7371f)
Am I configuring it wrong or does YOLOv8 not support rectangular input tensort at all?
P. S. `custom.yaml` is nothing special:
```
path: ./car-old
train: images/train
val: images/val
nc: 1  # number of classes
names: ['car']  # class names
```

**Comment by glenn-jocher** [comment]
@Shassk hello! It looks like you're on the right track with using the `rect` parameter for rectangular training in YOLOv8. However, the behavior you're experiencing seems unexpected. YOLOv8 does indeed support rectangular input tensors, and the `imgsz` parameter should allow you to specify the desired input resolution as a list `[width, height]`.

If you're observing a square input tensor despite setting `rect=True` and providing a rectangular `imgsz`, it's possible there might be an issue with the configuration or a bug. I would recommend double-checking the syntax in your YAML file and ensuring that the `imgsz` parameter is correctly specified in the training script.

If everything seems correct and the issue persists, please open an issue on the Ultralytics YOLOv8 GitHub repository with the details of your configuration and the unexpected behavior. This will allow the team to investigate further and provide assistance. Don't forget to include the version of the repository you're using and any relevant logs or error messages.

For more detailed guidance on training configurations, you can refer to our documentation at https://docs.ultralytics.com. Thank you for reaching out, and we appreciate your contribution to the YOLO community! 🚀

**Comment by Shassk** [comment]
Double checking logs I've foung this line:
`WARNING ⚠️ updating to 'imgsz=512'. 'train' and 'val' imgsz must be an integer, while 'predict' and 'export' imgsz may be a [h, w] list or an integer, i.e. 'yolo export imgsz=640,480' or 'yolo export imgsz=640'`.

<details open>
  <summary>Full log:</summary>

    Ultralytics YOLOv8.1.3 🚀 Python-3.10.12 torch-2.1.0+cu121 CUDA:0 (Tesla T4, 15102MiB)
    engine/trainer: task=detect, mode=train, model=yolov8n.yaml, data=/content/datasets/car/custom.yaml, epochs=20, time=None, patience=50, batch=56, imgsz=[512, 288], save=True, save_period=-1, cache=True, device=None, workers=8, project=None, name=train, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=True, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0, auto_augment=randaugment, erasing=0.4, crop_fraction=1.0, cfg=None, tracker=botsort.yaml, save_dir=runs/detect/train
    Downloading https://ultralytics.com/assets/Arial.ttf to '/root/.config/Ultralytics/Arial.ttf'...
    100%|██████████| 755k/755k [00:00<00:00, 118MB/s]
    Overriding model.yaml nc=80 with nc=1
    
                       from  n    params  module                                       arguments                     
      0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]                 
      1                  -1  1      4672  ultralytics.nn.modules.conv.Conv             [16, 32, 3, 2]                
      2                  -1  1      7360  ultralytics.nn.modules.block.C2f             [32, 32, 1, True]             
      3                  -1  1     18560  ultralytics.nn.modules.conv.Conv             [32, 64, 3, 2]                
      4                  -1  2     49664  ultralytics.nn.modules.block.C2f             [64, 64, 2, True]             
      5                  -1  1     73984  ultralytics.nn.modules.conv.Conv             [64, 128, 3, 2]               
      6                  -1  2    197632  ultralytics.nn.modules.block.C2f             [128, 128, 2, True]           
      7                  -1  1    295424  ultralytics.nn.modules.conv.Conv             [128, 256, 3, 2]              
      8                  -1  1    460288  ultralytics.nn.modules.block.C2f             [256, 256, 1, True]           
      9                  -1  1    164608  ultralytics.nn.modules.block.SPPF            [256, 256, 5]                 
     10                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
     11             [-1, 6]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
     12                  -1  1    148224  ultralytics.nn.modules.block.C2f             [384, 128, 1]                 
     13                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
     14             [-1, 4]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
     15                  -1  1     37248  ultralytics.nn.modules.block.C2f             [192, 64, 1]                  
     16                  -1  1     36992  ultralytics.nn.modules.conv.Conv             [64, 64, 3, 2]                
     17            [-1, 12]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
     18                  -1  1    123648  ultralytics.nn.modules.block.C2f             [192, 128, 1]                 
     19                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]              
     20             [-1, 9]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
     21                  -1  1    493056  ultralytics.nn.modules.block.C2f             [384, 256, 1]                 
     22        [15, 18, 21]  1    751507  ultralytics.nn.modules.head.Detect           [1, [64, 128, 256]]           
    YOLOv8n summary: 225 layers, 3011043 parameters, 3011027 gradients, 8.2 GFLOPs
    
    TensorBoard: Start with 'tensorboard --logdir runs/detect/train', view at http://localhost:6006/
    Freezing layer 'model.22.dfl.conv.weight'
    AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...
    Downloading https://github.com/ultralytics/assets/releases/download/v8.1.0/yolov8n.pt to 'yolov8n.pt'...
    100%|██████████| 6.23M/6.23M [00:00<00:00, 269MB/s]
    WARNING ⚠️ NMS time limit 0.550s exceeded
    AMP: checks passed ✅
    WARNING ⚠️ updating to 'imgsz=512'. 'train' and 'val' imgsz must be an integer, while 'predict' and 'export' imgsz may be a [h, w] list or an integer, i.e. 'yolo export imgsz=640,480' or 'yolo export imgsz=640'
    train: Scanning /content/datasets/car/labels/train... 189 images, 0 backgrounds, 1 corrupt: 100%|██████████| 189/189 [00:00<00:00, 1729.15it/s]train: WARNING ⚠️ /content/datasets/car/images/train/20658871.jpg: ignoring corrupt image/label: negative label values [ -0.0041494]
    train: New cache created: /content/datasets/car/labels/train.cache
    
    train: Caching images (0.1GB True): 100%|██████████| 188/188 [00:03<00:00, 60.32it/s]
    albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01), CLAHE(p=0.01, clip_limit=(1, 4.0), tile_grid_size=(8, 8))
    WARNING ⚠️ 'rect=True' is incompatible with DataLoader shuffle, setting shuffle=False
    val: Scanning /content/datasets/car/labels/val... 32 images, 0 backgrounds, 1 corrupt: 100%|██████████| 32/32 [00:00<00:00, 997.53it/s]val: WARNING ⚠️ /content/datasets/car/images/val/20008805.jpg: ignoring corrupt image/label: negative label values [ -0.0013793]
    val: New cache created: /content/datasets/car/labels/val.cache
    
    val: Caching images (0.0GB True): 100%|██████████| 31/31 [00:01<00:00, 25.74it/s]
    Plotting labels to runs/detect/train/labels.jpg... 
    optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
    optimizer: AdamW(lr=0.002, momentum=0.9) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0004375), 63 bias(decay=0.0)
    Image sizes 512 train, 512 val
    Using 2 dataloader workers
    Logging results to runs/detect/train
    Starting training for 20 epochs...
</details>

This matches [Train](https://docs.ultralytics.com/modes/train/#arguments) docs where only a single integer is stated:
| Key | Value | Description |
| --- | --- | --- |
| `imgsz` | `640` | size of input images as integer |

Unlike [Export](https://docs.ultralytics.com/modes/export/#arguments) where indeed 2 values can be passed:
| Key | Value | Description |
| --- | --- | --- |
| `imgsz` | `640` | image size as scalar or (h, w) list, i.e. (640, 480) |

Which is confusing. If it is not allowed during training then what happens when you set it in export?
P. S. An archive with a notebook and a small dataset to reproduce is attached.
[rect_minimal_reproducible_example.zip](https://github.com/ultralytics/ultralytics/files/13996745/rect_minimal_reproducible_example.zip)

**Comment by glenn-jocher** [comment]
Hello @Shassk, thank you for the detailed logs and the additional information. It appears that there is a discrepancy in the documentation regarding the `imgsz` parameter for training. Currently, YOLOv8 requires a single integer for the `imgsz` during training, which sets both the height and width of the input images to the same value, resulting in a square tensor.

For the `export` mode, you can specify a rectangular input size as a list `[height, width]`, which is useful for deployment scenarios where the model may need to process images of different aspect ratios.

I understand this can be confusing, and I appreciate you bringing this to our attention. We will review the documentation to ensure it accurately reflects the capabilities of the training mode and provide clearer guidance for users.

Regarding rectangular training, this is something we may consider for future updates. For now, please use a single integer for the `imgsz` during training. Thank you for your contribution to improving YOLOv8, and please continue to share any further questions or feedback you may have. 🌟

**Comment by Shassk** [comment]
Mo, @glenn-jocher, the square part is quite understandable, but I needed a confirmation of it for a current version, thanks for that. And we indeed have 16:9 input images from cameras.
But I still have 2 questions regarding resizing:
1. If I set rectangular input during export what will happen with other inputs and weights? Let's say I trained 512x512, exported resizing to 512x288 like I wanted, will additional parameters be just cut away? If so — what part of them? `(512 - 288) / 2` on each side? Or just first/last `(512 - 288)` of them?
2. Does resizing input affects sizes of input layers? Because even training a model in 320x320 resulted in only 0.1 MB difference (11.6 → 11.5) with pretty much identical performance. And the number of parameters have changed only for the last output layer. Am I right to assume keeping a larger size is benefitial since you are not losing any performance over smaller sizes but have more results in the end? Well, aside from memory usage, but that's a given.

**Comment by glenn-jocher** [comment]
Hello @Shassk,

1. When you export a model that was trained with a square input size (e.g., 512x512) to a rectangular input size (e.g., 512x288), the model's weights remain unchanged. The network architecture is designed to handle variable input sizes, and the weights are shared across different spatial dimensions. The resizing during export adjusts the input layer to accept the new aspect ratio, but the weights from training are preserved. No parameters are "cut away"; the model dynamically adapts to the new input size.

2. Resizing the input does affect the sizes of internal feature maps, but not the number of parameters in the network, except for layers directly connected to the input or output. The number of parameters is largely determined by the architecture and depth of the network, not the input size. Larger input sizes can indeed be beneficial as they may retain more information, potentially leading to better performance, provided that the increase in computational cost and memory usage is acceptable for your application.

I hope this clarifies your questions! 🚀

**Comment by torch-web** [comment]
I have a question. If rectangular training is used, the input image size is [512, 288]，If the input image size using rectangular training is [512, 288], but the actual input model size is [512, 512], and the exported onnx size is set to [512, 288], will accuracy be lost during actual detection?

**Comment by torch-web** [comment]
I have a question. If the input image size using rectangular training is [512, 288], but the actual input model size is [512, 512], and the exported onnx size is set to [512, 288], will accuracy be lost during actual detection?

**Comment by glenn-jocher** [comment]
@torch-web hey there! 😊

When you train with a square input (512x512) and export to a rectangular size (512x288), the model adapts without losing accuracy for the specified export size. It's designed to handle variable input sizes dynamically. Just ensure your training data is representative of the aspect ratio you plan to use during inference. Here's a quick example for exporting:

```python
from ultralytics import YOLO
model = YOLO('path/to/your/model.pt')
model.export(format='onnx', imgsz=[512, 288])
```

This way, your model is ready for detection with the intended aspect ratio. Happy detecting! 🚀
